upmc-enterprises/cda2fhir
# Feature/upmcfhir 228 jolt report

#### What does this PR do?

This is the  implementation of Jolt based transformation testing for CDA2FHIR allergy entries and the ability to document CCDA to FHIR transformations automatically from Jolt templates. Documentation is done in the form of csv files. It is tested on the sample files.

Currently only AllergyConcernAct CCDA entry template is implemented.

The CSV file consist of the following columns
1) Target: This is the field (or '.' joined fields) in Allergy Intolerance that is populated by the transform. All AllergyIntolerance fields are prefixed by AllergyIntolerance. AlerrgyConcernAct template also creates other resources and those are represented by the resource type in the Target (Practitioner, PractitionerRole) 
2) CCDA Source: This is primarily CCDA AllergyConcernAct template path that populates the target. It can also be a path starts with resources created within the transform. For example we use Allergy.reaction[].onset to populate lastOcurrance or Practitioner to populate recorder.reference
3) Format: This a function that formats the raw value from CCDA. Things like dateAdapter or valueSetAdapter. It is either a scalar function or a string function that reduces to a scalar (like max or selectFirst).
4) Link: Some objects comes up over and over in templates like CCDA ID object to FHIR Identifier or CCDA CD object to FHIR Codeable Concept transforms. These transforms are not shown explicitly. Instead a link column shows what the CCDA Source is left as a basic object which will be transformed by the transform in the link. These transforms have their own CSV files.
5) Condition 1, 2, 3, ...: These columns show the conditions for the transforms to happen. They are expressed as a CCDA Path + relation + value. relation can be 'isnull', 'isnotnull', 'equal' and 'notequal'. Value only exists for 'equal' and 'notequal' relations.

The outputs for AllergyIntolerance, ID and CD are in src/test/resources/gold/jolt-report. File names follow the input AllergyConcernAct.csv, ID.csv, CD.csv

#### Related JIRA tickets:

UPMCFHIR-227
UPMCFHIR-228

#### How should this be manually tested?

1) Looks at the AllergyConcernAct jolt transform template in src/test/resources/jolt/entry/AllergyConcernAct.json and make sure everything is being represented in the csv file.
2) Look at the java code and make sure the transform matches to the code.

#### Background/Context:

NA

#### New JIRA tickets for clinical review required?

NA
